### ISSUE TITLE
Confusion about upsampling results

### ISSUE THREAD

**Issue by Jaqen81** [aniqueakhtar/PointCloudUpsampling#1]
Hi, thanks for your great work! I've successfully configured the env and used the ckpt that you provided which is iter64000.pth(both 4x and 8x) to upsampling the kitti raw data.
My GT.pcd looks like this, contains 123531 points:
![GT](https://user-images.githubusercontent.com/46512334/169328506-8d0fb048-e92b-4f30-b2ed-f3628f47e4ad.png)
And this is input.pcd that after 8x downsampled has  15441 points:
![downsample](https://user-images.githubusercontent.com/46512334/169328961-c337312a-6e92-413f-a7c5-71418d087512.png)
8x upsampling pcd looks like this, 
![recon](https://user-images.githubusercontent.com/46512334/169329064-327d5623-6720-4378-a0de-054977e287b0.png)
which looks like each point has quite an obvious cube expanded, can I please ask if you have ever accounted for this problem? Thanks in advance!

**Comment by aniqueakhtar** [comment]
Hello,
So there are a number of issues with your approach:
You are taking already sparse data like the Kitti dataset and you are further downsampling it. And you are not downsampling it by 2x or 4x but 8x downsampling. That makes the resulting point cloud very sparse. I doubt the currently trained model of PU-Dense with 8x upsampling would be able to handle it because they were trained with a kernel_size (`ks`) variable of 5. You would need to increase that and retrain the model. `ks` variable determines the size of the last convolution kernel size and it has to be large to cater for sparser datasets.

I am also not sure if you pre-processed the KITTI dataset properly. If you are using a very sparse dataset, then the `k` value in the preprocessing (voxelization/quantization) would not work. Since creating an extremely sparse dataset means no overlaps, which would mean the convolutions would basically not work. You might need to change the `k` value to something like less than 20. If your dataset is very sparse, the convolutions would simply not overlap between points and hence you would not be able to apply convolutions to your data.

However, I would suggest not downsampling the KITTI dataset any further and simply using the input as it is with a `k` value between 30 and 100. If you want a more uniform output, I have seen a `k` value of 30 works really well during KITTI preprocessing. Also 4x upsampling of the KITTI dataset is giving really good results and 8x upsampling is also giving great results. You would have to play with the preprocessing variable `k`. 

If you do want to use a very sparse dataset, I am afraid you will have to retrain the network with a larger `ks` variable and sparser training dataset that is preprocessed by a smaller `k`.

**Comment by mdane-x** [comment]
Hi,

Thank you for the nice explanation. What `ks` kernel size would be appropriate? What's a possible range for the `ks` parameter? 3 - 20?

**Comment by aniqueakhtar** [comment]
I have typically found ks=5 to work well on most datasets.

**Comment by mdane-x** [comment]
Hi @aniqueakhtar ,

I retrained the model from scratch using sparse data (like you mentioned above) and used `ks` = 7 and `ks` = 15. In both cases I get the cubes that @Jaqen81 mentioned above.

What do you mean by pre-processing kitti? Why is it needed? 
I see your code downsamples automatically to generate ground truth, therefore I just pass point clouds.

Also do you normalize the data (e.g. 0 to 1)?

Thank you for the explanation.

**Comment by aniqueakhtar** [comment]
> What do you mean by pre-processing kitti? Why is it needed?

Kitti is not a voxelized dataset. You cannot use 3D convolutions / sparse convolutions without voxelizing the dataset first.
In the `Convert_Kitti.py` use a value of `k` variable between `30` and `100` to actually voxelize your point cloud. From my understanding, using a value of `k` as 50 gives good results. This is also why you are getting cubes. Please check the input tensor to the network. I am sure it has very few points because MinkowskiNet voxelized the data which results in losing most of the input points. Make sure the input tensor from the `data_loader` has the same number of points you want the network to process (you can print the input tensor shape).

Also, I would suggest downsampling Kitti by 8 times might not be a good idea. Nonetheless, please try playing with the "k" variable to voxelize the Kitti dataset and the "ks" variable to make sure that the kernel size is large enough to generate points in areas where you want.

We have trained our network on the ShapeNet dataset. To test the network on sparser datasets, it would be advisable to train our network on a sparser dataset too (one option would be to downsample the ShapeNet dataset).
